napari/utils/progress.py

```diff
@@ -150,3 +151,30 @@ def progrange(*args, **kwargs):
 
     """
     return progress(range(*args), **kwargs)
+
+
+class cancelable_progress(progress):
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        nest_under: Optional['progress'] = None,
+        cancel_callback: Optional[Callable] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        self.cancel_callback = cancel_callback
+        self.is_canceled = False
+
+        super().__init__(iterable, desc, total, nest_under, *args, **kwargs)
+
+    def __iter__(self) -> Iterator:
+        return takewhile(lambda _: not self.is_canceled, self.iterable)
```

Comment:
This leaves us unable to see progress bar updates in the GUI, and the bar doesn't disappear after cancellation. I think it should be:

```suggestion
        return takewhile(lambda _: not self.is_canceled, super().__iter__())
```